File: projects/grind-75/python/string_lc_questions/ransom_note.py
class Solution:
    @classmethod
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        ransom_count = {}
        magazine_count = {}
        for char in magazine:
            magazine_count[char] = magazine_count.get(char, 0) + 1
        for char in ransomNote:
            ransom_count[char] = ransom_count.get(char, 0) + 1
        for char, count in ransom_count.items():
            if magazine_count.get(char, 0) < count:
                return False
        return True

File: projects/grind-75/typescript/string/ransom-note.ts
export default function canConstruct(
  ransomNote: string,
  magazine: string,
): boolean {
  const ransomCount: { [char: string]: number } = {};
  const magazineCount: { [char: string]: number } = {};

  for (const c of ransomNote) {
    ransomCount[c] = (ransomCount[c] || 0) + 1;
  }
  for (const c of magazine) {
    magazineCount[c] = (magazineCount[c] || 0) + 1;
  }
  for (const c of ransomNote) {
    if ((magazineCount[c] || 0) < ransomCount[c]) {
      return false;
    }
  }
  return true;
}
